File: sdk/typescript/test/store.test.ts
import { describe, it, expect } from 'vitest';
import { pathToHash } from '../src/store.js';

describe('pathToHash', () => {
  it('converts a simple absolute path', () => {
    expect(pathToHash('/Users/keith/Github/k8/projects')).toBe('Users-keith-Github-k8-projects');
  });

  it('converts root path to empty string', () => {
    expect(pathToHash('/')).toBe('');
  });

  it('converts single segment path', () => {
    expect(pathToHash('/tmp')).toBe('tmp');
  });

  it('handles relative paths', () => {
    expect(pathToHash('relative/path')).toBe('relative-path');
  });

  it('truncates long paths and appends hash', () => {
    const segments = Array.from({ length: 20 }, () => 'abcdefghij');
    const longPath = '/' + segments.join('/');
    const result = pathToHash(longPath);

    expect(result.length).toBe(209);
    expect(result[200]).toBe('-');

    const suffix = result.slice(201);
    expect(suffix.length).toBe(8);
    expect(suffix).toMatch(/^[0-9a-f]{8}$/);
  });

  it('does not truncate paths at exactly 200 characters', () => {
    const base = 'a'.repeat(199);
    const path = '/' + base;
    const result = pathToHash(path);
    expect(result.length).toBeLessThanOrEqual(200);
    expect(result).not.toContain('-');
  });
});

File: sdk/elixir/test/direnv_config/store_test.exs
defmodule DirenvConfig.StoreTest do
  use ExUnit.Case, async: true

  alias DirenvConfig.Store

  describe "path_to_hash/1" do
    test "simple path" do
      assert Store.path_to_hash("/Users/keith/Github/k8/projects") ==
               "Users-keith-Github-k8-projects"
    end

    test "root path" do
      assert Store.path_to_hash("/") == ""
    end

    test "single segment" do
      assert Store.path_to_hash("/tmp") == "tmp"
    end

    test "truncation for long paths" do
      segments = List.duplicate("abcdefghij", 20)
      path = "/" <> Enum.join(segments, "/")
      hash = Store.path_to_hash(path)

      assert byte_size(hash) == 209
      assert binary_part(hash, 200, 1) == "-"

      suffix = binary_part(hash, 201, 8)
      assert Regex.match?(~r/^[0-9a-f]{8}$/, suffix)
    end
  end

  describe "state_dir/0" do
    test "uses XDG_STATE_HOME when set" do
      original = System.get_env("XDG_STATE_HOME")

      try do
        System.put_env("XDG_STATE_HOME", "/custom/state")
        assert Store.state_dir() == "/custom/state/direnv-config"
      after
        case original do
          nil -> System.delete_env("XDG_STATE_HOME")
          val -> System.put_env("XDG_STATE_HOME", val)
        end
      end
    end

    test "falls back to ~/.local/state/direnv-config" do
      original = System.get_env("XDG_STATE_HOME")

      try do
        System.delete_env("XDG_STATE_HOME")
        expected = Path.join([System.user_home!(), ".local", "state", "direnv-config"])
        assert Store.state_dir() == expected
      after
        if original, do: System.put_env("XDG_STATE_HOME", original)
      end
    end
  end
end
